src/platform/cmx/kernel.c

```diff
@@ -0,0 +1,271 @@
+void HAL_TIM_Base_MspDeInit(TIM_HandleTypeDef *htim)
+{
+    if(htim->Instance==TIM3)
+	{
+		/* Set the interrupt priority */
+		//HAL_NVIC_SetPriority(TIM3_IRQn,1,3);
+		/* Enable timer 3 interrupt */
+		HAL_NVIC_DisableIRQ(TIM3_IRQn);
+	}
+}
+
+/* The actual tim3 handler */
+struct pt_regs* timer_regs;
+void tim3irqhandler(struct pt_regs* regs)
+{
+	timer_regs=regs;
+	HAL_TIM_IRQHandler(&TIM3_Handler);
+}
+
+extern int timer_process(struct pt_regs *regs);
+
+/* The call-back function */
+void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim)
+{
+	if(htim==(&TIM3_Handler))
+	{
+		/* We are using one-shot, so disable the timer now */
+		HAL_TIM_Base_DeInit(&TIM3_Handler);
+		chal_timer_disable();
+		timer_process(timer_regs);
+	}
+}
+
+
+/* Systick is used to simulate the rdtsc. do not touch */
+void
+timer_init(void)
+{
+	HAL_SYSTICK_CLKSourceConfig(SYSTICK_CLKSOURCE_HCLK);
+	SysTick->CTRL|=SysTick_CTRL_TICKINT_Msk;
+	SysTick->LOAD=0xFFFFFF;/*1000;*/
+	SysTick->VAL=0x00;
+	NVIC_SetPriority(SysTick_IRQn,0x00);
+	SysTick->CTRL|=SysTick_CTRL_ENABLE_Msk;
+}
+
+void SysTick_Handler(void)
+{
+	rdtsc_sim++;
+}
+
+void __attribute__((optimize("O0"))) delay_us(u32 nus)
+{
+	u32 ticks;
+	u32 told,tnow,tcnt=0;
+	u32 reload=SysTick->LOAD;
+	ticks=nus*216;
+	told=SysTick->VAL;
+	while(1)
+	{
+		tnow=SysTick->VAL;
+		if(tnow!=told)
+		{
+			if(tnow<told)tcnt+=told-tnow;
+			else tcnt+=reload-tnow+told;
+			told=tnow;
+			if(tcnt>=ticks)break;
+		}
+	};
+}
+
+void
+GPIO_Init(void)
+{
+	/* We initialize the GPIOB 0,1, push-pull */
+	GPIO_InitTypeDef GPIO_Initure;
+	__HAL_RCC_GPIOB_CLK_ENABLE();
+	GPIO_Initure.Pin=GPIO_PIN_0|GPIO_PIN_1;
+	GPIO_Initure.Mode=GPIO_MODE_OUTPUT_PP;
+	GPIO_Initure.Pull=GPIO_PULLUP;
+	GPIO_Initure.Speed=GPIO_SPEED_HIGH;
+	HAL_GPIO_Init(GPIOB,&GPIO_Initure);
+	/* Turn off the LED */
+	HAL_GPIO_WritePin(GPIOB,GPIO_PIN_0,GPIO_PIN_SET);
+	HAL_GPIO_WritePin(GPIOB,GPIO_PIN_1,GPIO_PIN_SET);
+}
+
+void
+LED_set(void)
+{
+	HAL_GPIO_WritePin(GPIOB,GPIO_PIN_1,GPIO_PIN_RESET);
+	HAL_GPIO_WritePin(GPIOB,GPIO_PIN_0,GPIO_PIN_SET);
+}
+
+void
+LED_reset(void)
+{
+	HAL_GPIO_WritePin(GPIOB,GPIO_PIN_1,GPIO_PIN_SET);
+	HAL_GPIO_WritePin(GPIOB,GPIO_PIN_0,GPIO_PIN_RESET);
+}
+
+void
+main(void)
+{
+	/* The kernel image area */
+	timer_init();
+
+	SDRAM_Init();
+	LCD_Init();
+	GPIO_Init();
+
+	POINT_COLOR=RED;
+	LCD_Clear(GREEN);
+
+	kern_memory_setup();
+
+	chal_init();
+	cap_init();
+	ltbl_init();
+	retype_tbl_init();
+	comp_init();
+	thd_init();
+
+	/* PRY: Retype anything between mod_start and mod_end to user mem - This is for adding frames only */
+//	for(unsigned long i=0;i<0x4000;i+=1024)
+//	{
+//		assert(retypetbl_retype2user(&Memory_Used[i])==0);
+//	}
+
+	kern_boot_comp();
+	kern_boot_upcall();
+	/* should not get here... */
+	khalt();
+}
+
+void
+khalt(void)
+{
+	printk("Shutting down...\n");
+	while(1);
+}
+
+void
+printk(const char *fmt, ...)
+{
+
```

Comment:
I directly print out to the screen, so no printk needed currently. When I fix up the display buffer I can do printk with this.